test.sh

```diff
@@ -64,10 +97,107 @@ function run_test_coverage() {
   goveralls -v -coverprofile=gover.coverprofile -service=travis-pro
 }
 
+#
+# Main CLI Parser
+#
+USAGE="$(cat -- <<-EOM
+
+Usage:
+Boulder test suite CLI, intended to be run inside of a Docker container:
+
+  docker-compose run --use-aliases boulder ./$(basename "${0}") [OPTION]...
+
+With no options passed: runs standard battery of tests (lint, unit, and integation)
+    -l, --lints                           Adds lint to the list of tests to run
+    -u, --unit                            Adds unit to the list of tests to run
+    -d, --unit-test-filter <DIRECTORY>    Run unit tests for a specific directory
+    -e, --enable-unit-race                Enables -race flag for all unit tests
+    -n, --config-next                     Changes BOULDER_CONFIG_DIR from test/config to test/config-next
+    -c, --coverage                        Adds coverage to the list of tests to run
+    -i, --integration                     Adds integration to the list of tests to run
+    -s, --start-py                        Adds start (py) to the list of tests to run
+    -v, --gomod-vendor                    Adds gomod-vendor to the list of tests to run
+    -g, --generate                        Adds generate to the list of tests to run
+    -r, --rpm                             Adds rpm to the list of tests to run
+    -p, --show-integration-test-list      Outputs a list of the available integration tests
+    -f, --integration-filter <REGEX>      Run only those tests and examples matching the regular expression
```

Comment:
Let's make the long name here `--filter`, and apply it to both integration tests and unittests.

Reply:
Can you elaborate on this a bit?

Reply:
I think the idea is that being able to do
```./test.sh --unit --filter ./core/...```
and
```./test.sh --integration --filter TestSomething*```
is preferable to having to separate out
```./test.sh --unit --unit-dir-filter ./core/...```
and
```./test.sh --integration --integration-filter TestSomething*```.

But I recognize that this has a problem if you want to do
```./test.sh --unit --integration --filter ohnowhatgoeshere```

Reply:
`go test` supports a `-run` flag, which takes a regex. When we run integration tests, that how we pass the regex for the filter. It's also useful to be able to easily pass a `-run` flag to unittests, and it's convenient to use the same flag. So e.g. if you're running unittests and pass `--filter Foo`, test.sh would pass `-run Foo` to the `go test` command. That also lets us have a shorter longopt, since we don't have to specify that this is for "integration" only.

Reply:
I think we can make it an error to run both unit and integration with a filter. All the cases I can think of for filter involve running just one or the other.

Reply:
Good discussion here. I've gone ahead and implemented it the version that errors if you supply all three at the same time.